### PR TITLE
add remove_block_from_free_list

### PR DESCRIPTION
**Pull request by vasilyrud**
[https://media.giphy.com/media/rxiKQNnwMiMU/giphy.gif](https://media.giphy.com/media/rxiKQNnwMiMU/giphy.gif)

Changes:

1. added the function `remove_block_from_free_list`

2. added functionality for predecessor blocks since the function turned out to be impossible without it (i.e. slow and repetitive. but at least that means we now know why the doubly-linked list is necessary)

3. added asserts to the tests for placing block as well as for removing blocks (Quan, your helper test functions are super!)

4. I will work on the function `extend_heap` right now

@quanvuong

### CODE REVIEW

File: mm.c [hunk 1]
```diff
@@ -179,6 +179,37 @@ static void alloc_free_block(void *bp, size_t words) {
...
+	char **nearby_block;
+	int index;
+
+	// if largest block in free list set free list to next ptr
+	if (prev_block == NULL) {
+		index = find_free_list_index(GET_SIZE(bp));
+		main_free_list[index] = next_block;
+	} else { // if not largest block update pointer for prev block to next ptr
+		nearby_block = prev_block;
+		SET_PTR(GET_PTR_SUCC_FIELD(nearby_block), next_block);
```
Comment (quanvuong) [inline mm.c]: why don't you use prev_block directly here as arg to GET_PTR_SUCC_FIELD()

File: mm.c [hunk 2]
```diff
@@ -179,6 +179,37 @@ static void alloc_free_block(void *bp, size_t words) {
...
+	}
+
+	// next_block is not NULL, update the block to point to prev block
+	if (next_block != NULL) {
+		nearby_block = next_block;
+		SET_PTR(GET_PTR_PRED_FIELD(nearby_block), prev_block);
```
Comment (quanvuong) [inline mm.c]: same here

File: mm.c [hunk 3]
```diff
@@ -464,16 +500,27 @@ static void test_GET_PTR_SUCC_FIELD()
...
     assert(GET_SUCC(bp) == (*ptr_to_succ_field));
     printf("Test passed.\n\n");
     free(bp);
 }
 
+static void test_GET_PRED()
+{
+    printf("Test test_GET_PRED.\n");
+    char **bp = malloc(2*WORD_SIZE); // enough for hdr + pred field
+    char **ptr_to_pred_field = bp + HDR_SIZE; // bp should point to the pred field after this line
```
Comment (quanvuong) [inline mm.c]: for the comment, shouldn't it be "ptr_to_pred_field should point to the pred field after this line"

File: mm.c [hunk 4]
```diff
@@ -623,6 +678,52 @@ static void test_place_block_into_free_list()
...
 
+static void test_remove_block_from_free_list() {
+	printf("Test remove_block_from_free_list.\n");
+
+	char *free_list_backup[MAX_POWER + 1]; // backing up main_free_list to restore after test
+	for (int i = 0; i <= MAX_POWER; i++) {
+			free_list_backup[i] = main_free_list[i];
+	}
+
+	test_place_block_into_free_list_helper();
```
Comment (quanvuong) [inline mm.c]: should rename this helper, maybe to "initialize_free_lists_for_test" ?

**Comment by vasilyrud** [comment]
Also, I just realized that we didn't make use of the macro `GET_FREE_LIST_PTR` and `SET_FREE_LIST_PTR`. Will make use of those while revising as well.

**Comment by vasilyrud** [comment]
Revised

**Comment by vasilyrud** [comment]
Good to go?

**Comment by quanvuong** [comment]
yep LGTM

**Pull request merged**